### DISCUSSION TITLE
Assumptions of the minifier

### DISCUSSION THREAD

**Discussion by kdy1** [swc-project/swc#6074] [Ideas]
swc minifier analyzes the relation between variables, functions, and all declarations, and then it aborts if it cannot prove that it’s safe to optimize. This is correct, but requires lots of time and decreases the compression rate.

This is fine IMO, but I recently found that terser is not verifying it fully. I didn’t know it because I didn’t look at the compressor code of `terser`, but while reviewing [a PR](https://github.com/swc-project/swc/pull/6071), @Austaras told me that `terser` simply assumes that function declarations are not modified by side-effects of expressions. 

Related code: https://github.com/terser/terser/blob/4255fb3063ce1b41a5d5c9c48df443b5df0e797e/lib/compress/tighten-body.js#L887

There are some more similar assumptions.

`terser` **intentionally** assumes 

 - Function declarations are not modified by side effects.
 - Getters do not have side effects.
 - TDZ violation does not exist in the code.

I think we should collect these assumptions and write them on the documentation

**Comment by Austaras** [comment]
Not sure about 1 or 2, but 3 is definitely true

**Reply by jridgewell** [reply]
> Getters do not have side effects.

There's a setting that controls this. By default, it's `strict`, which makes terser recognize side-effects.
